sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/TLSEvaluationStage.java

```
@@ -0,0 +1,291 @@
+        this.certificateAuthorityClient = certificateAuthorityClient;
+        this.secretsClient = secretsClient;
+        this.keyPairGenerator = keyPairGenerator;
+    }
+
+    public static TLSEvaluationStage fromEnvironmentForService(String serviceName, String taskName) throws IOException, InvalidKeySpecException {
+
+        SchedulerFlags flags = SchedulerFlags.fromEnv();
+
+        KeyFactory keyFactory = null;
+        try {
+            keyFactory = KeyFactory.getInstance("RSA");
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
+        }
+
+        PemReader pemReader = new PemReader(new StringReader(flags.getServiceAccountPrivateKeyPEM()));
+        PrivateKey privateKey = keyFactory.generatePrivate(
+                new PKCS8EncodedKeySpec(pemReader.readPemObject().getContent()));
+
+        ServiceAccountIAMTokenProvider serviceAccountIAMTokenProvider = new ServiceAccountIAMTokenProvider.Builder()
+                .setIamUrl(URLHelper.fromUnchecked(DcosConstants.IAM_AUTH_URL))
+                .setUid(flags.getServiceAccountUid())
+                .setPrivateKey((RSAPrivateKey) privateKey)
+                .build();
+        TokenProvider tokenProvider = new CachedTokenProvider(serviceAccountIAMTokenProvider);
+
+        Executor executor = Executor.newInstance(
+                new HttpClientBuilder()
+                        .setTokenProvider(tokenProvider)
+                        .setLogger(LOGGER)
+                        .setRedirectStrategy(new LaxRedirectStrategy())
+                        .build());
+
+        CertificateAuthorityClient certificateAuthorityClient = new DefaultCAClient(executor);
+        SecretsClient secretsClient = new DefaultSecretsClient(executor);
+
+        KeyPairGenerator keyPairGenerator = null;
+        try {
+            keyPairGenerator = KeyPairGenerator.getInstance("RSA");
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
```

Stacktraces should be logged as errors and the method should return an exception, right?